dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/compilation/VisitorPTX.java

```diff
@@ -0,0 +1,75 @@
+package com.dat3m.dartagnan.program.processing.compilation;
+
+import com.dat3m.dartagnan.expression.ExprInterface;
+import com.dat3m.dartagnan.expression.IExpr;
+import com.dat3m.dartagnan.expression.IExprBin;
+import com.dat3m.dartagnan.program.Register;
+import com.dat3m.dartagnan.program.event.Tag;
+import com.dat3m.dartagnan.program.event.arch.ptx.RedOp;
+import com.dat3m.dartagnan.program.event.arch.ptx.AtomOp;
+import com.dat3m.dartagnan.program.event.core.Event;
+import com.dat3m.dartagnan.program.event.core.Load;
+import com.dat3m.dartagnan.program.event.core.rmw.RMWStore;
+
+import java.util.List;
+
+import static com.dat3m.dartagnan.program.event.EventFactory.newRMWLoad;
+import static com.dat3m.dartagnan.program.event.EventFactory.newRMWStore;
+import static com.dat3m.dartagnan.program.event.EventFactory.eventSequence;
+import static com.dat3m.dartagnan.program.event.EventFactory.newLocal;
+
+public class VisitorPTX extends VisitorBase {
+
+    protected VisitorPTX(boolean forceStart) {
+        super(forceStart);
+    }
+
+    @Override
+    public List<Event> visitPTXRMWFetchOp(AtomOp e) {
+        Register resultRegister = e.getResultRegister();
+        String mo = e.getMo();
+        IExpr address = e.getAddress();
+        ExprInterface value = e.getMemValue();
+        Register dummy = e.getThread().newRegister(resultRegister.getPrecision());
+        Load load = newRMWLoad(dummy, address, Tag.PTX.loadMO(mo));
+        RMWStore store = newRMWStore(load, address,
+                new IExprBin(dummy, e.getOp(), (IExpr) value), Tag.PTX.storeMO(mo));
+        for (String filter : e.getFilters()) {
+            load.addFilters(Tag.PTX.loadMO(filter));
+            load.addFilters(Tag.PTX.scopeMo(filter));
+            load.addFilters(Tag.PTX.proxyMo(filter));
+            store.addFilters(Tag.PTX.storeMO(filter));
+            store.addFilters(Tag.PTX.scopeMo(filter));
+            store.addFilters(Tag.PTX.proxyMo(filter));
+        }
```

Comment:
I don't think you need any of these

Reply:
Removing this part will break some atom tests.

Reply:
When you create a `Load(reg, address, mo)`, the `mo` tag is already added to the set of tags. This means that the tag you are added in lines 38 and 41 were already added. You can definitely remove those lines (both for Atom and Red events). 

What I guess you also need to do, is to add the tags related to the scope and the proxy. However iterating over all filters seems the wrong way to do it. You could simply do
```
if(e.is(Tag.PTX.SYS)) {
  load.add(Tag.PTX.SYS);
  store.add(Tag.PTX.SYS);
}
```
And the same for the other tags.

Reply:
Accepted

Reply:
You still have this
```
        for (String filter : e.getFilters()) {
            load.addFilters(Tag.PTX.loadMO(filter));
            store.addFilters(Tag.PTX.storeMO(filter));
        }
```
As I said, you don't need this because those tags are added because they are the `mo` when created the load and the store
```
        Load load = newRMWLoad(dummy, address, Tag.PTX.loadMO(e.getMo()));
        RMWStore store = newRMWStore(load, address,
                new IExprBin(dummy, e.getOp(), (IExpr) e.getMemValue()), Tag.PTX.storeMO(e.getMo()));
```

Reply:
Removed